src/server/qgsserverresponse.cpp

```diff
@@ -84,5 +80,7 @@ void QgsServerResponse::write( const QgsServerException &ex )
   setStatusCode( ex.responseCode() );
   setHeader( "Content-Type", responseFormat );
   write( ba );
-}
 
+  // log exception on server side too
+  QgsMessageLog::logMessage( ba, QStringLiteral( "Server" ), Qgis::Critical );
```

Comment:
Just wondering if "critical" is the right level here, maybe we should have two different levels, one for kind-of "expected" exceptions (the exceptions described in the various specifications of the services) and one for the kind-of "unhandled" exceptions, and reserve the "critical" level to the latter?

Reply:
Hi @elpaso,

For example, in case of `QgsServiceException( QStringLiteral( "UnknownError" ), ...)` (like in `qgswmsgetlegendgraphics.cpp`), we log a `CRITICAL` message (because it's an **UnknownError**). 

But in case of a `QgsBadRequestException( QStringLiteral( "LayerNotDefined" ), ... )`, we just raise a `WARNING` message considering that it's an expected error.

Am I understanding you correctly?

Reply:
@pblottiere yes, exactly, the idea is that an exception which is listed in the specs as a possible outcome of a bad call is less severe (less critical) than an exception that come from some internal checks within QGIS Server. 

Perhaps we should add the vendor parameters exceptions to the critical pool, like the ones from GETPRINT and such.

Reply:
I do agree wint @elpaso, those exception are usually legitimate as server response (error 400, 404 etc...) so they are not 'unexpected' and are the result of an error that has been handled correctly.

AMHA, this exception should be logged as  'request response'  -  and because they are legitimate http response they also will be logged the http level - soI would prefer the INFO level as not to increase too much the verbosity of the logging (INFO mode is already very verbose) .

For information, there is  a catch all for those exception  - and for unhandled exception -: https://github.com/qgis/QGIS/blob/master/src/server/qgsserver.cpp#L347.  This is mainly because code that throw such exception do not call `QgsServerResponse::write` directly, and may have logged there own debug info.

The only part where `QgsServerResponse::write` is called directly is in `QgsRequestHandler::setServiceException` because it is intended to be called in python and you cannot raise  C++ exception from python.

Furthemore I have noticed the catch all for unhandled exception (i.e http 500) - https://github.com/qgis/QGIS/blob/master/src/server/qgsserver.cpp#L351- return the reason in the body of the error response. I think that it is not appropriate for security reason (my bad) and while the ex.what() should be logged here has a CRITICAL error, the body of the sendError should only contain: 'Internal server Error'

Reply:
@elpaso @dmarteau I updated the code according to your suggestions. Let me know what you think :).

Reply:
@pblottiere  that's fine but sorry if I was not clear:  my remark about the 'catch all' was just informational:
 
Your original idea to log in the write() method is ok because you handle the case where the exception is set from `QgsRequestHandler::setServiceException`

Reply:
@elpaso What do you think about the changes?

Reply:
@pblottiere  :+1: thanks for working on this!